### PR TITLE
api: add user namespaces field to NodeRuntimeHandlerFeatures

### PR DESCRIPTION
**Pull request by sohankunkerkar**
Depended by:
https://github.com/kubernetes/kubernetes/pull/123303
https://github.com/kubernetes/website/pull/46953
https://github.com/kubernetes/kubernetes/pull/125259

#### What type of PR is this?
/kind feature
/kind api-change

#### What this PR does / why we need it:
We need this change to support the `ProcMountType` option.

#### Which issue(s) this PR fixes:

None

#### Does this PR introduce a user-facing change?

```release-note
Add UserNamespaces field to NodeRuntimeHandlerFeatures
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/4265
```

### CODE REVIEW

**Comment by haircommander** [comment]
/lgtm

from node side This mirrors RRO, where we use this for discovering whether a node supports user namespaces.

**Review by mrunalp — APPROVED** [review]

File: staging/src/k8s.io/api/core/v1/types.go [hunk 1]
```diff
@@ -6028,7 +6032,7 @@ type NodeStatus struct {
 	// +optional
 	Config *NodeConfigStatus `json:"config,omitempty" protobuf:"bytes,11,opt,name=config"`
 	// The available runtime handlers.
-	// +featureGate=RecursiveReadOnlyMounts
+	// +featureGate=RecursiveReadOnlyMounts,UserNamespacesSupport
```
Comment (thockin) [inline staging/src/k8s.io/api/core/v1/types.go:6035]: I think this should be on 2 lines

File: staging/src/k8s.io/api/core/v1/types.go [hunk 2]
```diff
@@ -5877,7 +5877,11 @@ type NodeRuntimeHandlerFeatures struct {
 	// +featureGate=RecursiveReadOnlyMounts
 	// +optional
 	RecursiveReadOnlyMounts *bool `json:"recursiveReadOnlyMounts,omitempty" protobuf:"varint,1,opt,name=recursiveReadOnlyMounts"`
-	// Reserved: UserNamespaces *bool (varint 2, for consistency with CRI API)
+	// UserNamespaces is set to true if the runtime handler supports UserNamespaces.
```
Comment (thockin) [inline staging/src/k8s.io/api/core/v1/types.go:5880]: "... if the runtime handler supports UserNamespaces, including idmapped mounting".

But do we want this to be SO specific?  Or is more possible to phrase more generically, like "...if the runtime handler supports UserNamespaces, including PersistentVolumes"

File: staging/src/k8s.io/api/core/v1/types.go (hunk 1 above)
Comment (sohankunkerkar) [reply staging/src/k8s.io/api/core/v1/types.go:6035]: @thockin Thanks for the suggestion. I will address this concern. I was looking for some docs here for the future reference. Any pointers?

File: staging/src/k8s.io/api/core/v1/types.go (hunk 2 above)
Comment (sohankunkerkar) [reply staging/src/k8s.io/api/core/v1/types.go:5880]: I think the latter sounds good to me. Let me go ahead and update the description.

**Comment by thockin** [comment]
From API POV this is pretty straight-forward, happy to approve when the rest is reviewed.

**Comment by haircommander** [comment]
/lgtm

still

File: pkg/kubelet/nodestatus/setters.go [hunk 3]
```diff
@@ -485,7 +485,7 @@ func GoRuntime() Setter {
 // RuntimeHandlers returns a Setter that sets RuntimeHandlers on the node.
 func RuntimeHandlers(fn func() []kubecontainer.RuntimeHandler) Setter {
 	return func(ctx context.Context, node *v1.Node) error {
-		if !utilfeature.DefaultFeatureGate.Enabled(features.RecursiveReadOnlyMounts) {
+		if !utilfeature.DefaultFeatureGate.Enabled(features.RecursiveReadOnlyMounts) && !utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport) {
```
Comment (kwilczynski) [inline pkg/kubelet/nodestatus/setters.go:488]: I wish this was done as a subsequent if-statement.

File: pkg/registry/core/node/strategy.go [hunk 4]
```diff
@@ -103,7 +103,7 @@ func dropDisabledFields(node *api.Node, oldNode *api.Node) {
 		node.Spec.ConfigSource = nil
 	}
 
-	if !utilfeature.DefaultFeatureGate.Enabled(features.RecursiveReadOnlyMounts) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.RecursiveReadOnlyMounts) && !utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport) {
```
Comment (kwilczynski) [inline pkg/registry/core/node/strategy.go:106]: Ditto. As a similar comment.

**Pull request merged**